File: leetcode2/2medium/최은서/Q1376.js
/**
 * @param {number} n
 * @param {number} headID
 * @param {number[]} manager
 * @param {number[]} informTime
 * @return {number}
 */
var numOfMinutes = function(n, headID, manager, informTime) {
    const tree = new Map();
    for (let i = 0; i < n; i++) {
        if (!tree.has(manager[i])) tree.set(manager[i], [])
        tree.get(manager[i]).push(i)
    }
    // 2 -> [0, 1, 3, 4, 5], -1 -> [2]
    const queue = [[headID, 0]]
    let maxTime = 0;
    while (queue.length) {
        const [id, time] = queue.shift();
        maxTime = Math.max(maxTime, time)
        if (tree.get(id)) {
            for (const nxt of tree.get(id)) {
                queue.push([nxt, time + informTime[id]])
            }
        }
    }
    return maxTime;
};

File: leetcode2/2medium/최원준/Q1376.java
package Leetcode;

/*
1. 아이디어 :


2. 시간복잡도 :
O( )

3. 자료구조/알고리즘 :

 */

import java.util.*;

public class Q1376 {
    class Solution {
        public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
            Map<Integer, List<Integer>> childs = new HashMap<>();
            for (int child=0; child<n; child++) {
                int par = manager[child];
                childs.putIfAbsent(par, new ArrayList<>());
                childs.get(par).add(child);
            }

            int ans = 0;

            Deque<int[]> deque = new ArrayDeque<>();
            deque.add(new int[]{headID, 0});

            while (!deque.isEmpty()) {
                int[] c = deque.pollFirst();
                int curr = c[0], time = c[1];
                ans = Math.max(ans, time);

                for (int child : childs.getOrDefault(curr, new ArrayList<>())) {
                    deque.add(new int[]{child, time + informTime[curr]});
                }
            }
            return ans;
        }
    }
}
